scipy/_lib/_util.py

```diff
@@ -966,3 +966,85 @@ def _dict_formatter(d, n=0, mplus=1, sorter=None):
+
+    Array arguments may have any number of core dimensions (typically 0,
+    1, or 2) and any broadcastable batch shapes. There may be any
+    number of array outputs of any number of dimensions. Assumptions
+    right now - which are satisfied by all functions of interest in `linalg` -
+    are that all array inputs are consecutive keyword or positional arguments,
+    and that the wrapped function returns either a single array or a tuple of
+    arrays. It's only as general as it needs to be right now - it can be extended.
+
+    Parameters
+    ----------
+    *argdefs : tuple of (str, int)
+        Definitions of array arguments: the keyword name of the argument, and
+        the number of core dimensions.
+
+    Example:
+    --------
+    `linalg.eig` accepts two matrices as the first two arguments `a` and `b` and
+    returns one array or a tuple of arrays  depending on the values of other
+    positional or keyword arguments. To generate a wrapper that applies the function
+    over batches of matrices:
+
+    >>> _apply_over_batch(('a', 2), ('b', 2))
+    """
+    names, ndims = list(zip(*argdefs))
+    n_arrays = len(names)
+
+    def decorator(f):
+        @functools.wraps(f)
+        def wrapper(*args, **kwargs):
+            args = list(args)
+
+            # Ensure all arrays in `arrays`, other arguments in `other_args`/`kwargs`
+            arrays, other_args = args[:n_arrays], args[n_arrays:]
+            for i, name in enumerate(names):
+                if name in kwargs:
+                    if i + 1 <= len(args):
+                        raise ValueError(f'{f.__name__}() got multiple values '
+                                         f'for argument `{name}`.')
+                    else:
+                        arrays.append(kwargs.pop(name))
+
+            # Determine batch shapes
+            batch_shapes = []
+            for i, (array, ndim) in enumerate(zip(arrays, ndims)):
+                array = None if array is None else np.asarray(array)
+                shape = () if array is None else array.shape
+                arrays[i] = array
+                batch_shapes.append(shape[:-ndim] if ndim > 0 else shape)
+
+            # Early exit if call is not batched
+            if not any(batch_shapes):
+                return f(*arrays, *other_args, **kwargs)
+
+            # Determine broadcasted batch shape
+            batch_shape = np.broadcast_shapes(*batch_shapes)  # Gives OK error message
+
+            # Main loop
+            results = []
+            for index in np.ndindex(batch_shape):
+                result = f(*(array[*index] for array in arrays), *other_args, **kwargs)
+                # Distinguish between single output and multiple outputs
+                # Could improve that are define the wrapper with that information
+                result = (result,) if not isinstance(result, tuple) else result
+                results.append(result)
```

Comment:
I did not perform any measurements but this step, the `list(zip(...))` and `np.stack` below will make this less performant and memory hungry since you will be holding all the nD-slices and the growing `result` in memory simultaneously until it finishes and you will be pushing a growing array to the garbage collector which can start affecting performance during the computation.

A preallocation of `result` and placing the results to each slice of the `result` is a less wasteful solution in my opinion. I did a similar thing in `expm` as follows (basically looping over the slices in an ndarray). 

https://github.com/scipy/scipy/blob/04ec412b9d990cf2bfa3704e769b6a4d185c799a/scipy/linalg/_matfuncs.py#L312-L313

If it is not nD but just 2D, the loop spins only once.

Reply:
Still not ideal since these are much less wasteful to do in Cython or C because LAPACK calls then performed inplace but until then we can live with it.

Reply:
Yes, I considered pre-allocating. It would have made it a bit more complicated because I would have to call the function once to figure out the result type, pre-allocate, then do the loop.

Also, we don't want this to be *too* efficient, right? Need some motivation to come back and improve the efficiency with compiled code, right? : )

Anyway, if you want me to make it a little more complex for that performance bump, I will. I was just trying to start as simple as possible.

Reply:
Certainly, but it shouldn't be too bad in my opinion compared to user is calling eig in a loop and doing it themselves. Especially when we have a chance to do it relatively OK fashion.

Reply:
OK, I'll add this in the next commit.

Reply:
Ah, the other complication is that the result dtype of `eig` is data-dependent.

```python3
import numpy as np
from scipy import linalg
rng = np.random.default_rng(4932982394)
m = 10

A = rng.random((m, m))
evals, evecs = linalg.eig(A)
print(evals.dtype, evecs.dtype)

A += A.T + np.eye(m) * m
evals, evecs = linalg.eig(A)
print(evals.dtype, evecs.dtype)
```
Perhaps this is someting else to fix, but I think it's a separate issue.  I would prefer not to *also* check each iteration whether we need to reallocate the array with a new dtype. Can we leave it for a follow-up?

Reply:
@ev-br Can we continue the discussion about pre-allocating memory in this thread?

Again, I considered pre-allocating, but opted against it when implementing this originally because I don't think it is worth the complexity, especially when this is going to be replaced anyway.

Here is a simplified case using `solve` (because there is only one output) to compare the two options.

```python3
import numpy as np
rng = np.random.default_rng(4578345834932546)
n, m = 10000, 100
A = rng.random((n, m, m))
b = rng.random(m)

def preallocate(A, b):
    shape = A.shape[:-1]
    out0 = np.linalg.solve(A[0], b)
    out = np.empty(shape, dtype=out0.dtype)
    out[0] = out0
    for i, Ai in enumerate(A[1:]):
        out[i+1] = np.linalg.solve(Ai, b)
    return out

def stack(A, b):
    return np.stack([np.linalg.solve(Ai, b) for Ai in A])

res = preallocate(A, b)  # 939 ms ± 13.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
ref = stack(A, b)  # 955 ms ± 11.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Because the time is dominated by the looping and execution of the function, preallocation is not substantially faster. 

The only difference is a momentary factor of 2 on peak memory consumption. (Before the `stack`, the stacking option requires on average half the memory, and after the `stack`, the list of arrays can be garbage collected.)

Note that this `preallocate` version is actually _simpler_ than it would need to be, considering that the output dtype of `eig` is data-dependent: I am only checking the dtype of the output of the first slice, but really, it would need to be checked in each iteration. If the dtype goes from real to complex, we'd have to reallocate, move the existing results over, and pick up where we left off.

Can this PR focus on the interface, documentation, and correctness; and if someone wants to squeeze out that factor of two (before all this code is replaced anyway), that can be done separately?

Reply:
I think there are two things at play here, First, for large enough core dimensions, the linear algebra big-O complexity dominates; Second, indeed, when looping in python, pre-allocation does not buy much.
To see the perf budget, add a `np.linalg.solve` comparison: for `m=100` your python loop versions are essentially equivalent to numpy, while for `m=10` there's a factor of 5 or so:

```
In [39]: import numpy as np
    ...: rng = np.random.default_rng(4578345834932546)
    ...: n, m = 10000, 10
    ...: A = rng.random((n, m, m))
    ...: b = rng.random(m)
    ...: 
    ...: def preallocate(A, b):
    ...:     shape = A.shape[:-1]
    ...:     out0 = np.linalg.solve(A[0], b)
    ...:     out = np.empty(shape, dtype=out0.dtype)
    ...:     out[0] = out0
    ...:     for i, Ai in enumerate(A[1:]):
    ...:         out[i+1] = np.linalg.solve(Ai, b)
    ...:     return out
    ...: 
    ...: def stack(A, b):
    ...:     return np.stack([np.linalg.solve(Ai, b) for Ai in A])
    ...: 
    ...: %timeit np.linalg.solve(A, b)
    ...: %timeit preallocate(A, b)  # 939 ms ± 13.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
    ...: %timeit stack(A, b)  # 955 ms ± 11.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
13.9 ms ± 215 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
73.1 ms ± 265 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
73.5 ms ± 303 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Conclusion: as long as looping is in python, extra complexity is not worth it.